File: src/Simulation.java
import java.util.ArrayList;

/*
 * Copyright 2016 Mark Injerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class Simulation {
	/** Number of simulations per second. */
	static final int TICK_RATE = 100;
	/** Indicates if simulation and rendering is paused. */
	private static boolean isPaused;
	/** Indicates if gameplay has started. */
	private static boolean isStarted;
	private static long simulationTime = 0;
	private static final ArrayList<GameStateListener> listeners = new ArrayList<GameStateListener>();

	interface GameStateListener {
		void onGameStartStateChanged(boolean started);
		void onGamePauseStateChanged(boolean paused);
	}

	static void addGameStateListener(GameStateListener listener) {
		Simulation.listeners.add(listener);
	}

	static boolean isStarted() {
		return isStarted;
	}

	static boolean isPaused() {
		return isPaused;
	}

	static void setStarted(boolean start) {
		if (start == isStarted) return;
		isStarted = start;
		for (GameStateListener listener : listeners) {
			listener.onGameStartStateChanged(start);
		}
	}

	static void setPaused(boolean pause) {
		if (pause == isPaused) return;
		isPaused = pause;
		for (GameStateListener listener : listeners) {
			listener.onGamePauseStateChanged(pause);
		}
	}

	/** Runs one simulation cycle. */
	static void simulate() {
		if (isPaused) return;
		boolean isClient = MultiplayerManager.getInstance().isClient();
		Ship localShip = ShipManager.getLocalShip();
		simulationTime += 1000 / TICK_RATE;
		if (!localShip.isDestroyed()) {
			localShip.calculateMotion();
		}
		synchronized (Asteroid.getAsteroids()) {
			for (Asteroid a : Asteroid.getAsteroids()) {
				a.calculateMotion();
			}
		}
		if (isClient) {
			synchronized (localShip.getBullets()) {
				for (int i = localShip.getBullets().size() - 1; i >= 0; i--) {
					if (localShip.getBullets().get(i).isDestroyed()) {
						localShip.getBullets().remove(i);
					}
				}
			}
		} else {
			for (Ship ship : ShipManager.getShips()) {
				for (Bullet b : ship.getBullets()) {
					b.calculateMotion();
				}
			}
			checkCollisions();
		}
		MultiplayerManager.getInstance().sendShipUpdate();
	}

	private static void checkCollisions() {
		if (MultiplayerManager.getInstance().isClient()) return;
		Ship localShip = ShipManager.getLocalShip();
		for (Ship ship : ShipManager.getShips()) {
			// Ship vs. ship
			if (ship != localShip && localShip.isContacting(ship)) {
				localShip.collide(ship);
			}
			// Ship vs. asteroid
			synchronized (Asteroid.getAsteroids()) {
				for (int i = Asteroid.getAsteroids().size() - 1; i >= 0; i--) {
					Asteroid a = Asteroid.getAsteroids().get(i);
					if (ship.isContacting(a)) {
						ship.collide(a);
					}
				}
			}
			synchronized (ship.getBullets()) {
				for (int i = ship.getBullets().size() - 1; i >= 0; i--) {
					Bullet bullet = ship.getBullets().get(i);
					// Local bullet vs. remote ship
					if (ship == localShip) {
						for (Ship remoteShip : ShipManager.getShips()) {
							if (remoteShip == localShip) continue;
							if (bullet.isContacting(remoteShip)) {
								bullet.collide(remoteShip);
								break;
							}
						}
					}
					// Remote bullet vs. local ship
					if (ship != localShip && bullet.isContacting(localShip)) {
						bullet.collide(localShip);
						break;
					}
					// Bullet vs. asteroid
					synchronized (Asteroid.getAsteroids()) {
						for (int n = Asteroid.getAsteroids().size() - 1; n >= 0; n--) {
							Asteroid a = Asteroid.getAsteroids().get(n);
							if (bullet.isContacting(a)) {
								bullet.collide(a);
								bullet.setHitAsteroidSize(a.getSize());
								break;
							}
						}
					}
					updateScore(ship, bullet);
					if (bullet.isDestroyed()) {
						ship.getBullets().remove(i);
					}
				}
			}
		}
	}

	private static void updateScore(Ship ship, Bullet b) {
		Asteroid.Size hitAsteroidSize = b.getHitAsteroidSize();
		if (hitAsteroidSize != null) {
			ship.addScore(hitAsteroidSize.getScoreValue());
			b.setHitAsteroidSize(null);
		}
	}

	/**
	 * @return Time in milliseconds, based on simulation cycles (1000 / {@link #TICK_RATE} ms per cycle), that the simulation has been running (unpaused).
	 */
	static long getSimulationTime() {
		return simulationTime;
	}
}

File: src/Simulation.kt
/*
 * Copyright 2016 Mark Injerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.util.*

object Simulation {
	/** Number of simulations per second.  */
	const val TICK_RATE = 100

	/** Indicates if simulation and rendering is paused.  */
	private var isPaused = false

	/** Indicates if gameplay has started.  */
	var isStarted = false
		set(start) {
			if (start == isStarted) return
			field = start
			for (listener in listeners) {
				listener.onGameStartStateChanged(start)
			}
		}

	/**
	 * @return Time in milliseconds, based on simulation cycles (1000 / [.TICK_RATE] ms per cycle), that the simulation has been running (unpaused).
	 */
	var simulationTime: Long = 0
		private set
	private val listeners = ArrayList<GameStateListener>()
	fun addGameStateListener(listener: GameStateListener) {
		listeners.add(listener)
	}

	fun isPaused(): Boolean {
		return isPaused
	}

	fun setPaused(pause: Boolean) {
		if (pause == isPaused) return
		isPaused = pause
		for (listener in listeners) {
			listener.onGamePauseStateChanged(pause)
		}
	}

	/** Runs one simulation cycle.  */
	fun simulate() {
		if (isPaused) return
		val isClient: Boolean = MultiplayerManager.instance.isClient
		val localShip = ShipManager.localShip
		simulationTime += (1000 / TICK_RATE).toLong()
		if (!localShip.isDestroyed) {
			localShip.calculateMotion()
		}
		synchronized(Asteroid.asteroids) {
			for (a in Asteroid.asteroids) {
				a.calculateMotion()
			}
		}
		if (isClient) {
			synchronized(localShip.bullets) {
				localShip.bullets.removeIf { it.isDestroyed }
			}
		} else {
			for (ship in ShipManager.ships) {
				if (ship == null) continue
				for (b in ship.bullets) {
					b.calculateMotion()
				}
			}
			checkCollisions()
		}
		MultiplayerManager.instance.sendShipUpdate()
	}

	private fun checkCollisions() {
		if (MultiplayerManager.instance.isClient) return
		val localShip = ShipManager.localShip
		for (ship in ShipManager.ships) {
			if (ship == null) continue
			// Ship vs. ship
			if (ship != localShip && localShip.isContacting(ship)) {
				localShip.collide(ship)
			}
			// Ship vs. asteroid
			synchronized(Asteroid.asteroids) {
				for (i in Asteroid.asteroids.indices.reversed()) {
					val a: Asteroid = Asteroid.asteroids[i]
					if (ship.isContacting(a)) {
						ship.collide(a)
					}
				}
			}
			synchronized(ship.bullets) {
				for ((i, bullet) in ship.bullets.withIndex().reversed()) {
					// Local bullet vs. remote ship
					if (ship == localShip) {
						for (remoteShip in ShipManager.ships) {
							if (remoteShip == null || remoteShip == localShip) continue
							if (bullet.isContacting(remoteShip)) {
								bullet.collide(remoteShip)
								break
							}
						}
					}
					// Remote bullet vs. local ship
					if (ship != localShip && bullet.isContacting(localShip)) {
						bullet.collide(localShip)
						break
					}
					// Bullet vs. asteroid
					synchronized(Asteroid.asteroids) {
						for (n in Asteroid.asteroids.indices.reversed()) {
							val a: Asteroid = Asteroid.asteroids[n]
							if (bullet.isContacting(a)) {
								bullet.collide(a)
								bullet.hitAsteroidSize = a.size
								break
							}
						}
					}
					updateScore(ship, bullet)
					if (bullet.isDestroyed) {
						ship.bullets.removeAt(i)
					}
				}
			}
		}
	}

	private fun updateScore(ship: Ship, b: Bullet) {
		b.hitAsteroidSize?.run {
			ship.addScore(scoreValue)
			b.hitAsteroidSize = null
		}
	}

	interface GameStateListener {
		fun onGameStartStateChanged(started: Boolean)
		fun onGamePauseStateChanged(paused: Boolean)
	}
}
